contracts/cw4-stake/src/contract.rs

```diff
@@ -67,37 +65,88 @@ pub fn handle(
 pub fn handle_bond(
     deps: DepsMut,
     env: Env,
-    info: MessageInfo,
+    amount: Balance,
+    sender: &HumanAddr,
 ) -> Result<HandleResponse, ContractError> {
     let cfg = CONFIG.load(deps.storage)?;
 
     // ensure the sent denom was proper
     // NOTE: those clones are not needed (if we move denom, we return early),
     // but the compiler cannot see that (yet...)
-    let sent = match info.sent_funds.len() {
-        0 => Err(ContractError::NoFunds {}),
-        1 => {
-            if info.sent_funds[0].denom == cfg.denom {
-                Ok(info.sent_funds[0].amount)
-            } else {
-                Err(ContractError::MissingDenom(cfg.denom.clone()))
+    let sent_funds = match amount {
+        Balance::Native(balance) => {
+            GenericBalance {
+                native: balance.0,
```

Comment:
I think this matching can be done a lot simpler.

Here is the check for a cw20 whitelist from escrow: https://github.com/CosmWasm/cosmwasm-plus/blob/master/contracts/cw20-escrow/src/contract.rs#L127-L132 (this is a bit easier as we only have one denom to check here)

I would suggest this code for checking the balance: https://github.com/CosmWasm/cosmwasm-plus/blob/master/packages/cw0/src/payment.rs#L13-L36 but I see it needs MessageInfo (which we don't have here). You could make and export another function `must_pay_fund()` that does the body, such that we have:

```rust
pub fn must_pay(info: &MessageInfo, denom: &str) -> Result<Uint128, PaymentError> {
    must_pay_funds(&info.sent_funds, denom)
}
```